## Treehelper functionalities

In [43]:
import sys
sys.path.append('..')
from ontology_handling import TreeHelper

In [44]:
ontohelper_dharani = TreeHelper(ontoname='dharani')

In [45]:
# ontohelper_dharani.print_tree()

In [46]:
len(ontohelper_dharani)

496

### search

In [47]:
ontohelper_dharani.search('orbital neocortex')

[SearchResult(matched_name='orbital cortex', score=90.32258064516128, ontoid=21)]

In [48]:
ontohelper_dharani.search('ca1',partial=True)

[SearchResult(matched_name='ca1 field of hippocampus', score=100.0, ontoid=386),
 SearchResult(matched_name='subplate zone, ca1 field', score=100.0, ontoid=420),
 SearchResult(matched_name='intermediate zone, ca1 field', score=100.0, ontoid=421)]

### subtrees (groups)

In [49]:
ontohelper_dharani.groups

['HPF',
 'AMY_BN',
 'Mig',
 'HY',
 'TH',
 'MB',
 'HB',
 'BS',
 'CB',
 'dev',
 'ft',
 'Vs',
 'Ctx']

In [50]:
ontohelper_dharani.print_subtree('BS')

[lvl] id (acronym) name
---------------------
 [0] 440 (BS) Brainstem
   [1] 192 (MB) Midbrain
     [2] 203 (Prt) Pretectum
       [3] 322 (APT) Anterior pretectal nucleus
       [3] 323 (MPT) Medial pretectal nucleus
       [3] 324 (PPT) Posterior pretectal nucleus
       [3] 325 (NOT) Nucleus of the optic tract
       [3] 326 (OPT) Olivary pretectal nucleus
     [2] 193 (Tct) Tectum
       [3] 194 (SC) Superior colliculus
       [3] 195 (IC) Inferior colliculus
         [4] 318 (ICc) Inferior colliculus, central nucleus
         [4] 319 (ICe) Inferior colliculus, external nucleus
         [4] 385 (ICd) Inferior colliculus, dorsal nucleus
     [2] 196 (Tgt) Tegmentum
       [3] 320 (PBN) Parabigeminal nucleus
       [3] 321 (NPC) Nucleus of the posterior commissure
       [3] 197 (SN) Substantia nigra
         [4] 198 (SNc) Substantia nigra pars compacta
         [4] 199 (SNr) Substantia nigra pars reticulata
       [3] 200 (R) Red nucleus
       [3] 285 (PAG) Periaqueductal gray
    

In [51]:
ontohelper_dharani.ontoids_by_group['BS']

[440, 192, 204, 210, 448, 445, 446, 447, 460, 461, 462, 463]

many ids are missed - this is because 'HB'(204) is a group by itself, and also a child of 'BS'. Similarly 'MB'(192)

### Special functions

These make sense only for Dharani

In [52]:
logicalareas = ontohelper_dharani.get_ids_of_cortical_areas()

In [53]:
logicalareas

defaultdict(list,
            {'-FCTx': [13, 14, 15, 16, 17, 18, 19],
             '-ORB': [22, 23, 24, 25, 26, 27, 28],
             '-PAR': [30, 31, 32, 33, 34, 35, 36],
             '-OCC': [39, 40, 41, 42, 43, 44, 45],
             '-TEMP': [48, 49, 50, 51, 52, 53, 54],
             '-ENT': [57, 58, 59, 60, 61, 62, 63],
             '-INS': [500, 501, 502, 503, 504, 505, 506],
             '-CING': [121, 122, 123, 124, 125, 126, 127]})

In [54]:
bylev = ontohelper_dharani.get_ids_by_level(1)

In [55]:
bylev

[2,
 440,
 215,
 317,
 367,
 368,
 369,
 370,
 371,
 158,
 442,
 1023,
 1022,
 1015,
 1021,
 1019,
 1018,
 1016,
 1020,
 1024,
 1000,
 1001,
 1002,
 1003,
 1004,
 1005,
 1006,
 1025,
 1007,
 1008,
 1009,
 1010,
 1013,
 1014,
 1026,
 1027,
 1028,
 455,
 468,
 469,
 466,
 467,
 232,
 235,
 243,
 244,
 415,
 417,
 337,
 338,
 252,
 237,
 238,
 251,
 334,
 414,
 248,
 431,
 343,
 327,
 341,
 342,
 311,
 346,
 347,
 457,
 456,
 250,
 458,
 253,
 225,
 226,
 227,
 228,
 443,
 229,
 230]

Ancestor ids returned by `get_ancestor_ids` are ordered from root to the immediate parent

In [56]:
ancestorids=ontohelper_dharani.get_ancestor_ids(22)
ancestorids

[1, 2, 3, 4, 21]

In [57]:
ontohelper_dharani.onto_lookup[22]

NodeRecord(acronym='SGL-ORB', name='Subpial granular layer, orbital cortex', color_hex_triplet='#b8a98a', level=5, parentid=21, numchildren=0)

In [58]:
layeredareas=ontohelper_dharani.get_ids_of_layered_areas()

In [59]:
layeredareas

defaultdict(list,
            {'SGL-': [13,
              22,
              30,
              39,
              48,
              57,
              67,
              75,
              86,
              94,
              102,
              110,
              500,
              121],
             'MZ-': [14,
              23,
              31,
              40,
              49,
              58,
              68,
              76,
              87,
              95,
              103,
              111,
              501,
              122],
             'CP-': [15,
              24,
              32,
              41,
              50,
              59,
              69,
              77,
              88,
              96,
              104,
              112,
              502,
              123],
             'SP-': [16,
              25,
              33,
              42,
              51,
              60,
              70,
              78,
              420,
              422,


## Advanced 

### checking the ontology json itself

The json might have useful fields like `definition`, `type`

In [60]:
nd = ontohelper_dharani._get_node_by_ontoid(21)

In [61]:
ontohelper_dharani._get_node_data(nd)

{'id': 21,
 'name': 'Orbital cortex',
 'acronym': 'ORB',
 'definition (description)': 'Note: may include only the orbital gyrus. Both terms (orbito-frontal ctx and orbital gyrus) appear in coronal and saggital plates, but only the orbital gyrus is listed in the glossary. The orbital sulcus seems to develop at a later g.w.',
 'type (gray matter/fiber tract/CNS cavity/developmental/other)': 'G',
 'parent_structure_id': 4,
 'color_hex_triplet': 'CFE4C2',
 'text': 'ORB : Orbital cortex'}

In [62]:
def show_node(node,lev = 0):
    if lev > 1:
        return
    
    for k,v in node.items():
        
        if k!='children':
            print("".join(["\t"]*lev), k, ":", v)
        else:
            for ch in node['children']:
                show_node(ch,lev+1)
        
show_node(nd)

 id : 21
 name : Orbital cortex
 acronym : ORB
 definition (description) : Note: may include only the orbital gyrus. Both terms (orbito-frontal ctx and orbital gyrus) appear in coronal and saggital plates, but only the orbital gyrus is listed in the glossary. The orbital sulcus seems to develop at a later g.w.
 type (gray matter/fiber tract/CNS cavity/developmental/other) : G
 parent_structure_id : 4
 color_hex_triplet : CFE4C2
	 id : 22
	 name : Subpial granular layer, orbital cortex
	 acronym : SGL-ORB
	 definition (description) : Needs a definition
	 type (gray matter/fiber tract/CNS cavity/developmental/other) : D
	 parent_structure_id : 21
	 color_hex_triplet : b8a98a
	 text : SGL-ORB : Subpial granular layer, orbital cortex
	 id : 23
	 name : Marginal zone, orbital cortex
	 acronym : MZ-ORB
	 definition (description) : See Kostovic et al. 2018
	 type (gray matter/fiber tract/CNS cavity/developmental/other) : D
	 parent_structure_id : 21
	 color_hex_triplet : de895e
	 text : MZ-OR

some checks

In [63]:
ontohelper_dharani.get_group_by_acronym('ORB')

'Ctx'

In [64]:
ontohelper_dharani.ontoids_by_group['MB']

[185,
 407,
 408,
 192,
 203,
 322,
 323,
 324,
 325,
 326,
 193,
 194,
 195,
 318,
 319,
 385,
 196,
 320,
 321,
 197,
 198,
 199,
 200,
 285,
 444,
 403,
 404,
 263,
 262,
 261,
 201,
 428,
 449,
 283,
 450,
 451,
 452,
 454,
 453,
 377]

Some ontoids might be ungrouped (not a part of any defined group)

In [65]:
ungroupedids = []
for id,rec in ontohelper_dharani.onto_lookup.items():
    if ontohelper_dharani.get_group_by_acronym(rec.acronym) is None:
        ungroupedids.append(id)

In [66]:
[ontohelper_dharani.onto_lookup[id] for id in ungroupedids]

[NodeRecord(acronym='Br', name='Brain', color_hex_triplet='#F6FAF4', level=0, parentid=0, numchildren=3),
 NodeRecord(acronym='FB', name='Forebrain', color_hex_triplet='#EFF6EA', level=1, parentid=1, numchildren=2),
 NodeRecord(acronym='Tel', name='Telencephalon', color_hex_triplet='#EFF6EA', level=2, parentid=2, numchildren=3),
 NodeRecord(acronym='Di', name='Diencephalon', color_hex_triplet='#FFF3F3', level=2, parentid=2, numchildren=2),
 NodeRecord(acronym='SP', name='Spinal cord', color_hex_triplet='#C0C1C2', level=0, parentid=0, numchildren=6),
 NodeRecord(acronym='DGSC', name='Dorsal gray of the spinal cord', color_hex_triplet='#AFB0B1', level=1, parentid=231, numchildren=0),
 NodeRecord(acronym='DH', name='Dorsal horn', color_hex_triplet='#AFB0B1', level=1, parentid=231, numchildren=0),
 NodeRecord(acronym='IH', name='Intermediate gray of the spinal cord', color_hex_triplet='#B1B2B3', level=1, parentid=231, numchildren=0),
 NodeRecord(acronym='CG', name='Central gray of the spin

### Fuzzy match into Allen ontology

In [67]:
ontohelper_allen = TreeHelper('allen_devhuman')

In [68]:
len(ontohelper_allen)

3315

In [69]:

def nameupdate(name):
    expansions = {
        'SG':'subpial granular layer',
        'MZ': 'marginal zone',
        'CP': 'cortical plate',
        'SP': 'subplate zone',
        'IZ': 'intermediate zone',
        'SZ': 'subventricular zone',
        'VZ': 'ventricular zone',
        'Hipp': 'hippocampus',
        # 'GE': 'ganglionic eminence',
    }
    if ' in ' in name:
        for acr,exp in expansions.items():
            name = name.replace(acr,exp)
        name = name.replace(' in ', ', ')
        
    if 'neocortex' in name:
        name = name.replace('neocortex','cortex')
    return name

updated_allen_dict = {k:nameupdate(v.name) for k,v in ontohelper_allen.onto_lookup.items()}

ontohelper_allen.search_dict = updated_allen_dict


In [70]:
ontohelper_allen.search('subiculum')

[SearchResult(matched_name='subiculum', score=100.0, ontoid=10302)]

In [71]:
from collections import defaultdict
from tqdm import tqdm

matches = defaultdict(list)

for dh_id,rec in ontohelper_dharani.onto_lookup.items():
    query = rec.name.lower()# .replace(',', ' in')
    ret = ontohelper_allen.search(query)
    
    matches[dh_id].extend(ret)

    if len(matches[dh_id])>1:
        print(query,matches[dh_id])

forebrain [SearchResult(matched_name='forebrain (prosencephalon)', score=100.0, ontoid=10156), SearchResult(matched_name='gray matter of forebrain', score=100.0, ontoid=10157), SearchResult(matched_name='transient structures of forebrain', score=100.0, ontoid=10506), SearchResult(matched_name='basal forebrain', score=100.0, ontoid=10349), SearchResult(matched_name='white matter of forebrain', score=100.0, ontoid=10557)]
frontal cortex [SearchResult(matched_name='frontal neocortex', score=90.32258064516128, ontoid=10161), SearchResult(matched_name='prefrontal cortex', score=90.32258064516128, ontoid=10172)]
marginal zone, frontal cortex [SearchResult(matched_name='marginal zone', score=100.0, ontoid=10508), SearchResult(matched_name='prefrontal cortex', score=90.32258064516128, ontoid=10172)]
subplate zone, frontal cortex [SearchResult(matched_name='subplate zone', score=100.0, ontoid=10522), SearchResult(matched_name='prefrontal cortex', score=90.32258064516128, ontoid=10172)]
subventr

In [72]:
for k,v in matches.items():
    name = ontohelper_dharani.onto_lookup[k].name
    if len(v)==1:
        pass
        # print('*',k,name,v)
    else:
        print(k,name,'---', v)
    

2 Forebrain --- [SearchResult(matched_name='forebrain (prosencephalon)', score=100.0, ontoid=10156), SearchResult(matched_name='gray matter of forebrain', score=100.0, ontoid=10157), SearchResult(matched_name='transient structures of forebrain', score=100.0, ontoid=10506), SearchResult(matched_name='basal forebrain', score=100.0, ontoid=10349), SearchResult(matched_name='white matter of forebrain', score=100.0, ontoid=10557)]
12 Frontal cortex --- [SearchResult(matched_name='frontal neocortex', score=90.32258064516128, ontoid=10161), SearchResult(matched_name='prefrontal cortex', score=90.32258064516128, ontoid=10172)]
14 Marginal zone, frontal cortex --- [SearchResult(matched_name='marginal zone', score=100.0, ontoid=10508), SearchResult(matched_name='prefrontal cortex', score=90.32258064516128, ontoid=10172)]
15 Cortical plate, frontal cortex --- []
16 Subplate zone, frontal cortex --- [SearchResult(matched_name='subplate zone', score=100.0, ontoid=10522), SearchResult(matched_name='

In [73]:
ontohelper_allen.search('pontine',partial=True)

[SearchResult(matched_name='prefrontopontine fibers', score=100.0, ontoid=12042),
 SearchResult(matched_name='frontopontine fibers', score=100.0, ontoid=12048),
 SearchResult(matched_name='parietopontine fibers', score=100.0, ontoid=12055),
 SearchResult(matched_name='occipitopontine fibers', score=100.0, ontoid=12056),
 SearchResult(matched_name='temporopontine fibers', score=100.0, ontoid=12061)]

In [74]:
for ontoid, rec in ontohelper_allen.onto_lookup.items():
    if 'field' in rec.name.lower():
        print(ontoid, rec)

10462 NodeRecord(acronym='FF', name='nucleus of the field of Forel', color_hex_triplet='#B67BB4', level=6, parentid=10461, numchildren=0)
15550 NodeRecord(acronym='CTF', name='central tegmental field', color_hex_triplet='#6CA9BF', level=5, parentid=12277, numchildren=0)
12289 NodeRecord(acronym='RRF', name='retrorubral field', color_hex_triplet='#5298AE', level=5, parentid=12277, numchildren=0)


In [75]:
ontohelper_allen.onto_lookup[11737]

NodeRecord(acronym='CPhipc', name='CP in caudal hippocampus', color_hex_triplet='#5D457F', level=7, parentid=11735, numchildren=0)

In [76]:
ontohelper_allen.onto_lookup[11735]

NodeRecord(acronym='CPhip', name='CP in hippocampal proper', color_hex_triplet='#5D457F', level=6, parentid=10520, numchildren=2)

In [77]:
ontohelper_dharani.onto_lookup[450]

NodeRecord(acronym='IPNdl', name='Interpeduncular nucleu, dorsolateral part', color_hex_triplet='#8EBDCD', level=5, parentid=283, numchildren=0)